### PR TITLE
Add derivation for ocs-url

### PR DESCRIPTION
**Pull request by SohamG**
Signed-off-by: Soham S Gumaste <sgumas2@uic.edu>

###### Description of changes

Add `ocs-url` to support one click installs from DE store pages like kde-store.com

~~Add .desktop file for handling ocs:// URI~~ The build produces the needed desktop file for handling ocs:// URI

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [??] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [X] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by SohamG** [comment]
I cant get editorconfig to work on my emacs or vs code. Please could someone tell me how I could fix the error?

File: nixos/doc/manual/release-notes/rl-2211.section.md [hunk 1]
```diff
@@ -118,6 +118,8 @@ Use `configure.packages` instead.
 
 - Add udev rules for the Teensy family of microcontrollers.
 
+- Add derivation for ocs-url with support for ocs:// URI.
```
Comment (trofi) [inline nixos/doc/manual/release-notes/rl-2211.section.md:121]: I think news items are written for user-facing changes and major changes. Most packages don't get mentioned in a NEWS file. I would suggest omitting NEWS updates.

File: pkgs/tools/misc/ocs-url/default.nix [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  version = "3.1.0";
+
+  meta.description = "Open Collaboration System for use with DE store websites";
+  meta.license = lib.licenses.gpl3Only;
+  meta.maintainers = [ lib.maintainers.SohamG ];
+  meta.platforms = lib.platforms.linux;
+  srcs = [
+    (fetchgit {
+      url = "https://www.opencode.net/dfn2/ocs-url.git";
+      rev = "refs/tags/release-3.1.0";
```
Comment (trofi) [inline pkgs/tools/misc/ocs-url/default.nix:14]: [untested] It's customary to use tags without full qualifier. While at it you can reuse version from package version:

```suggestion
      rev = "release-${version}";
```

File: pkgs/tools/misc/ocs-url/default.nix [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+      sha256 = "RvbkcSj8iUAHAEOyETwfH+3XnCCY/p8XM8LgVrZxrws=";
+    })
+
+    (fetchgit {
+      url = "https://github.com/akiraohgaki/qtil";
+      rev = "refs/tags/v0.4.0";
+      sha256 = "XRSp0F7ggfkof1RNAnQU3+O9DcXDy81VR7NakITOXrw=";
+    })
+  ];
```
Comment (trofi) [inline pkgs/tools/misc/ocs-url/default.nix:23]: Does the source need to be present in the same unit? Is it feasible to split it into another derivation?

File: pkgs/tools/misc/ocs-url/default.nix [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+  sourceRoot = "ocs-url";
+  dontWrapQtApps = true;
+
+  buildInputs = with libsForQt5.qt5;
+    [ qtbase qtsvg qtquickcontrols qmake ] ++ [ git ];
```
Comment (trofi) [inline pkgs/tools/misc/ocs-url/default.nix:29]: `git` should probably go to `nativeBuildInputs`.

Comment (trofi) [inline pkgs/tools/misc/ocs-url/default.nix:29]: `qt5` package usually get packages slightly differently and hide under qt5 namespace: via `libsForQt5.callPackage`.

File: pkgs/tools/misc/ocs-url/default.nix [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+  buildPhase = ''
+    source $stdenv/setup
```
Comment (trofi) [inline pkgs/tools/misc/ocs-url/default.nix:32]: `source $stdenv/setup` should already be present at build execution. Why do you need it?

File: pkgs/tools/misc/ocs-url/default.nix [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    runHook preBuild
```
Comment (trofi) [inline pkgs/tools/misc/ocs-url/default.nix:33]: I think you could make `buildPhase` a default if you moved things around:

1. Move to `postPatch` (or `postUnpack`) source fiddling:
```
    mkdir -p ./lib/qtil
    cp -r ../qtil/* lib/qtil/
```

2. Move to `configurePhase` (or `postConfigure`): `qmake PREFIX=$out`.

As a benefit you would get `make` ran with parallelism enabled by default.

File: pkgs/tools/misc/ocs-url/default.nix [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    mkdir -p ./lib/qtil
+    cp -r ../qtil/* lib/qtil/
+    qmake PREFIX=$out
+    make
+    runHook postBuild
+  '';
+  installPhase = "make install";
+
+  distPhase = ''
```
Comment (trofi) [inline pkgs/tools/misc/ocs-url/default.nix:42]: I would sugegst `postInstall` phase for it.

**Review by trofi — CHANGES_REQUESTED** [review]

File: pkgs/tools/misc/ocs-url/default.nix (hunk 5 above)
Comment (SohamG) [reply pkgs/tools/misc/ocs-url/default.nix:32]: Apologies, I'm new to Nix packaging; I thought since I was overriding buildPhase I needed to source that

**Comment by SohamG** [comment]
@trofi Thanks a lot for the review. I'll get to implementing the changes.

**Review by trofi — CHANGES_REQUESTED** [review]
Well done! I have a few small nits to complain about.

**Review by trofi — APPROVED** [review]
Perfect! LGTM

**Comment by trofi** [comment]
One more thing which is possibly related to Qt wrapping. Does the app start for you as is? Does it need KDE or something similar? Here is my failed attempt:

```
$ nix-build -A ocs-url
/nix/store/2b41jn5ldbrp4ish3lgpp5x5f8chs3bv-ocs-url
$ LANG=C ./result/bin/ocs-url
qt.qpa.plugin: Could not find the Qt platform plugin "xcb" in ""
This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.

Aborted (core dumped)
```

**Comment by trofi** [comment]
The following change gets something that runs for me and shows that I typed invalid URL:

```diff
--- a/pkgs/tools/misc/ocs-url/default.nix
+++ b/pkgs/tools/misc/ocs-url/default.nix
@@ -19,9 +19,8 @@ stdenv.mkDerivation rec {
   ];

   sourceRoot = "ocs-url";
-  dontWrapQtApps = true;

-  buildInputs = with libsForQt5.qt5; [ qtbase qtsvg qtquickcontrols qmake ];
+  buildInputs = with libsForQt5.qt5; [ qtbase qtsvg qtquickcontrols qmake wrapQtAppsHook ];

   # We are NOT in $sourceRoot here
   postUnpack = ''
```

**Comment by trofi** [comment]
Works for me as is now. Thank you!

**Pull request merged**